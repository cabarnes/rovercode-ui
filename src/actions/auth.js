// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
