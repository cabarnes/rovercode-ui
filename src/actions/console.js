// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
