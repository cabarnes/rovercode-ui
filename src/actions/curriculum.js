// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
