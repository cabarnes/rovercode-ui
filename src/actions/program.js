// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
