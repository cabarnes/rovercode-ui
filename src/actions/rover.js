// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SCAN = 'SCAN';
export const SCAN_FULFILLED = `${SCAN}_FULFILLED`;
export const SCAN_REJECTED = `${SCAN}_REJECTED`;
export const CONNECT_ROVER = 'CONNECT_ROVER';
export const CONNECT_ROVER_FULFILLED = `${CONNECT_ROVER}_FULFILLED`;
export const CONNECT_ROVER_REJECTED = `${CONNECT_ROVER}_REJECTED`;
export const SEND_ROVER = 'SEND_ROVER';
export const SEND_ROVER_FULFILLED = `${SEND_ROVER}_FULFILLED`;
export const SEND_ROVER_REJECTED = `${SEND_ROVER}_REJECTED`;
export const DISCONNECT_ROVER = 'DISCONNECT_ROVER';

const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
const UART_TX_CHARACTERISTIC_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
const UART_RX_CHARACTERISTIC_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

// action creators
export const scan = () => ({
  type: SCAN,
  payload: navigator.bluetooth.requestDevice({
    filters: [{ namePrefix: 'BBC micro:bit' }],
    optionalServices: [UART_SERVICE_UUID],
  }),
});

export const connect = (device, onMessage) => ({
  type: CONNECT_ROVER,
  payload: device.gatt.connect()
    .then(server => server.getPrimaryService(UART_SERVICE_UUID))
    .then(service => Promise.all([
      service.getCharacteristic(UART_RX_CHARACTERISTIC_UUID),
      service.getCharacteristic(UART_TX_CHARACTERISTIC_UUID)
        .then((characteristic) => {
          characteristic.startNotifications();
          characteristic.addEventListener(
            'characteristicvaluechanged',
            onMessage,
          );
          return characteristic;
        }),
    ])),
});

export const send = (channel, message) => ({
  type: SEND_ROVER,
  payload: channel.writeValue(message),
});

export const disconnect = device => ({
  type: DISCONNECT_ROVER,
  payload: device.gatt.disconnect(),
});
