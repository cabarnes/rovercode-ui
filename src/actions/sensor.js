// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});
