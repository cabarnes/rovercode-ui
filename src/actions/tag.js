// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_TAGS = 'FETCH_TAGS';
export const FETCH_TAGS_PENDING = `${FETCH_TAGS}_PENDING`;
export const FETCH_TAGS_FULFILLED = `${FETCH_TAGS}_FULFILLED`;
export const FETCH_TAGS_REJECTED = `${FETCH_TAGS}_REJECTED`;

// action creators
export const fetchTags = (xhrOptions) => ({
  type: FETCH_TAGS,
  payload: axios.get('/api/v1/tags/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
