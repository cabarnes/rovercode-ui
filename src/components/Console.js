import React, { Component } from 'react';
import { Button } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';

import { clear as actionClear } from '@/actions/console';

const mapStateToProps = ({ console }) => ({ console });
const mapDispatchToProps = (dispatch) => ({
  clear: () => dispatch(actionClear()),
});

class Console extends Component {
  constructor(props) {
    super(props);

    this.bottomRef = React.createRef();
  }

  componentDidUpdate() {
    this.bottomRef.current.scrollIntoView({ behavior: 'smooth' });
  }

  handleClear = () => {
    const { clear } = this.props;

    clear();
  }

  render() {
    const { console } = this.props;

    /* eslint-disable react/no-array-index-key */
    return (
      <>
        <div style={{ height: '200px', overflow: 'scroll' }}>
          {
            console.messages.map((message, index) => (
              <p key={index}>
                {`>> ${message}`}
              </p>
            ))
          }
          <div ref={this.bottomRef} />
        </div>
        <Button primary onClick={this.handleClear} style={{ marginTop: '10px' }}>
          <FormattedMessage
            id="app.console.clear"
            description="Button label to clear the console"
            defaultMessage="Clear"
          />
        </Button>
      </>
    );
  }
}

Console.propTypes = {
  console: PropTypes.shape({
    messages: PropTypes.array.isRequired,
  }).isRequired,
  clear: PropTypes.func.isRequired,
};

export default hot(module)(connect(mapStateToProps, mapDispatchToProps)(Console));
