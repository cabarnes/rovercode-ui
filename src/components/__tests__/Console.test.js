import React from 'react';
import { Button } from '@material-ui/core';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import { clear } from '@/actions/console';
import Console from '../Console';

describe('The Console component', () => {
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore({
      console: {
        messages: ['first', 'second'],
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const wrapper = mountWithIntl(<Console store={store} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays messages', () => {
    const wrapper = shallow(<Console store={store} />).dive().dive();

    expect(wrapper.find('p').length).toBe(2);
    expect(wrapper.find('p').at(0).text()).toBe('>> first');
    expect(wrapper.find('p').at(1).text()).toBe('>> second');
  });

  test('clears messages', () => {
    const wrapper = mountWithIntl(<Console store={store} />);

    wrapper.find(Button).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(clear());
  });

  test('scrolls to the bottom', () => {
    const mockScroll = jest.fn();
    const wrapper = shallow(<Console store={store} />).dive().dive();
    wrapper.instance().bottomRef = {
      current: {
        scrollIntoView: mockScroll,
      },
    };
    wrapper.instance().componentDidUpdate();

    expect(mockScroll).toHaveBeenCalled();
  });
});
