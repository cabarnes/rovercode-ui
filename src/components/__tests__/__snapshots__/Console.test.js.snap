// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Console component renders on the page with no errors 1`] = `
<Connect(Console)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Console
    clear={[Function]}
    console={
      Object {
        "messages": Array [
          "first",
          "second",
        ],
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <div
      style={
        Object {
          "height": "200px",
          "overflow": "scroll",
        }
      }
    >
      <p
        key="0"
      >
        &gt;&gt; first
      </p>
      <p
        key="1"
      >
        &gt;&gt; second
      </p>
      <div />
    </div>
    <Button
      as="button"
      onClick={[Function]}
      primary={true}
      style={
        Object {
          "marginTop": "10px",
        }
      }
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui primary button"
              style="margin-top: 10px;"
            >
              Clear
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui primary button"
                style="margin-top: 10px;"
              >
                Clear
              </button>,
            }
          }
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            style={
              Object {
                "marginTop": "10px",
              }
            }
          >
            <FormattedMessage
              defaultMessage="Clear"
              id="app.console.clear"
              values={Object {}}
            >
              Clear
            </FormattedMessage>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
  </Console>
</Connect(Console)>
`;
